Print CFN template of a load-balanced web application

efekarakus opened the pull request:
Print to `stdout` the CFN template for an application!

Parent story: #229 
The next step is to implement the functionality for `--output-dir`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

sonofachamp commented on internal/pkg/cli/app_package.go:
```diff
@@ -107,7 +115,19 @@ func (opts *PackageAppOpts) Validate() error {
 // Execute prints the CloudFormation template of the application for the environment.
 func (opts *PackageAppOpts) Execute() error {
-	return nil
+	project := viper.GetString(projectFlag)
```
Should there be a `--project`/`-p` flag either on the `app` sub-command or even root command where we bind the viper value to that flag? I think that should help remove these `viper.GetString(projectFlag)` throughout the codebase. I can't think of too many situations where a project name will not need to be in context.

efekarakus replied:
Ok this is what I'm thinking, y'all tell me if you like it:

1. We'll move the `bindProjectName()` function to `init` function in the `internal/cli` pkg.
2. All the `Opts` structs will have a private `projectName` field that's set to `viper.GetString("project")` in their constructor.

This way it will allow us to write tests without having the set the field with viper.

I don't think I see a con w/ this approach.

sonofachamp replied:
Could we bind the value to a `GlobalOpts` struct and embed that in sub-command `Opts`? I'm trying to think of a way to avoid the duplicate definitions so that if we add another "global option" it will just be available to sub-commands that rely on that structure.

efekarakus replied:
Good idea! Done :)

sonofachamp commented on internal/pkg/cli/app_package.go:
```diff
@@ -179,9 +221,9 @@ func BuildAppPackageCmd() *cobra.Command {
+	cmd.Flags().StringVarP(&opts.AppName, "name", "n", opts.AppName, "Name of the application.")
+	cmd.Flags().StringVarP(&opts.EnvName, "env", "e", opts.EnvName, "Name of the environment.")
+	cmd.Flags().StringVar(&opts.Tag, "tag", opts.Tag, `Optional. The application's image tag. Defaults to your latest git commit's hash.`)
+	cmd.Flags().StringVar(&opts.OutputDir, "output-dir", opts.OutputDir, "Optional. Writes the stack template and template configuration to a directory.")
```
Just for clarity would these ever be expected to bind values other than zero-values?

efekarakus replied:
Ah the reason why I made this change is that `NewPackageAppOpts` sets the `Tag` value to the commit-id.

If I gave the default value of `""` then it overrides the commitID :(

I'll add a comment explaining this

kohidave commented on internal/pkg/workspace/workspace.go:
```diff
@@ -247,7 +249,7 @@ func (ws *Workspace) WriteManifest(manifestBlob []byte, applicationName string)
 		return "", err
 	}
-	manifestFileName := fmt.Sprintf("%s%s", applicationName, manifestFileSuffix)
+	manifestFileName := fmt.Sprintf(FmtManifestFileName, applicationName)
```
Perhaps we should have a ManifestPath function

efekarakus commented:
I merged the change so that I can make progress on the `--output-dir` flag, I believe I addressed all the comments. But please let me know by commenting here if there is anything I should change in my next PR
